### PR TITLE
Adding COD and WOD dice system

### PR DESCRIPTION
**Pull request by novahorizon**
CoD (Chronicles of Darkness) and WoD (World of Darkness) both use a d10 system and sums the number of dice that roll above a certain value.

**CoD** counts the number of dice that equal or exceed 8, and always explode on 10, but may explode on 8 or 9 as well.

**WoD** counts the number of dice that equal or exceed a difficulty rating provided by the GM, always explode on 10, and fail (subtract) on 1.

The systems can be used by issuing the roll command with the dice pool, "cod" or "wod" as the delimitator, and the explode value/difficulty rating respectively.

For example:

**CoD:**
 Dice pool of 7d10 with "9-again" would be 
/roll 7cod9
 
 **WoD:**
Dice pool of 7d10 with difficulty 6 would be
/roll 7wod6

Reworked dice system into a couple classes to hopefully make future changes more intuitive.

added [highlighting] to certain dice results

### CODE REVIEW

**Comment by kreicer** [comment]
Hello, mate.

First of all, I wanna say thank you for the interest to this bot. This is really matter for me and very inspired.

I just return from vacation and didn't had time to explore your changes deeply yet. I have a question: 7cod9 and 7wod6 will be the same as 7d10/exp:9/suc:8 and 7d10/exp/suc:6/fail if we will add 2 new postfixes and mechanic to use more than 1 postfix per dice (was planned in 2.1.0)?

I asking cos I would prefer keep the main functional of the bot clear from game systems (fate is exception but on first implementation I tried make it as postfix also, current implementation - hack).

I see around 3 ways:
1) Add new postfixes "success" and "fail". And mechanic to use multiple postfixes.
- suc - success - converts each dice roll result to 1 (on "value" or higher) or 0, default = edge
- f - fail - converts each dice roll result to -1 (on "value" or lower) or 1, default = 1
  
  
2) Add new game system specific postfixes. Like cod and wod. Example: 7d10/cod:9 and 7d10/wod:6
- cod -  explode on value (default = edge) and converts each dice roll result to 1 (on 8 or higher) or 0
- wod - explode on 10 and and converts each dice roll result to 1 (on value or higher), -1 on 1 and 0 on another results
  
  
3) Move code related to cod and wod systems to separate cog. This allow make all even simple with CMDs like "/cod 7 9" and "/wod 7 6" or maybe "/darkness 7cod9".

What do you think about it?

**Comment by novahorizon** [comment]
Ah. I understand the desire to keep the system clear. I didn't understand the cog system well enough to try and implement something like that myself.

As part of this PR completely rewrites the rolling mechanic, I believe option 1 would be the least compatible solution. 

Option 2 could be something that simply becomes an alias in the check_dice_dict() method and easy to rewrite for. (This was not the case in the initial PR, but I've updated it quite a lot during your vacation. I hope you went somewhere nice and had a good time.)

Option 3 I imagine is also quite compatible. I would probably go with the /darkness command personally to reduce command clutter.


--

That said, now that the PR includes the rewrite for the "DiceBucket" class and "Dice" class, I think Fate could be rewritten into a postfix to generate the new type as well.


I wrote those two classes to hopefully make it easier to implement new postfixes and new game systems. I believe turning postfix into a list of tuples ("postfix", value) would be doable within the DiceBucket class with little effort.

File: cogs/roll.py
Comment (novahorizon) [inline cogs/roll.py:1]: Somehow I broke this in the process of changing the code around and had to rewrite it to be more lenient.

**Comment by kreicer** [comment]
Hey! 
Sorry for my disappear - I just had duty to release 2.1 as soon as possible. Now I have time to integrate your changes into main code, if you still interested.

**Comment by novahorizon** [comment]
I am still interested, yes. There have been many changes in your code however, so it may take a while for me to resolve any conflicts between our branches. There are some fundamental differences in our branches at this point as well. For us to merge, I would like your advice on how we should proceed.

**Comment by kreicer** [comment]
Sure.
I will look on conflicts asap and write my thinks.

**Comment by kreicer** [comment]
Hey.
I did looked at code.

Few questions:
- did you test execution speed of new classes?
- "added [highlighting] to certain dice results" - mb you have screenshot?

Right now we surely can integrate cod and wod postfixes to current schema. About classes - I bad at OOP. So, I should run stable version of your code changes to understand how it works - will do it this weekend.

**Comment by kreicer** [comment]
I'm done with code inspection.
I really like idea with class for dice and some other ideas. I just cant resolve code conflicts between our versions. Too many changes. 

I think we can add changes piece by piece. Lets keep for this MR only new postfixes: its model, checks and add it to current roll flow.

**Comment by kreicer** [comment]
Hey, mate.
Will you add changes into code?

**Pull request closed without merging**